File: histomicstk/preprocessing/color_deconvolution/separate_stains_macenko_pca.py
"""Placeholder."""
import numpy

import histomicstk.utils as utils

from . import _linalg as linalg
from .complement_stain_matrix import complement_stain_matrix


def separate_stains_macenko_pca(
        im_sda, minimum_magnitude=16, min_angle_percentile=0.01,
        max_angle_percentile=0.99, mask_out=None):
    """Compute the stain matrix for color deconvolution with the Macenko method.

    For a two-stain image or matrix in SDA space, this method works by
    computing a best-fit plane with PCA, wherein it selects the stain
    vectors as percentiles in the "angle distribution" in that plane.

    Parameters
    ----------
    im_sda : array_like
        Image (MxNx3) or matrix (3xN) in SDA space for which to compute the
        stain matrix.
    minimum_magnitude : float
        The magnitude below which vectors will be excluded from the computation
        of the angle distribution.

        The default is based on the paper value of 0.15, adjusted for our
        method of calculating SDA, thus 0.15 * 255 * log(10)/log(255)
    min_angle_percentile : float
        The smaller percentile of one of the vectors to pick from the angle
        distribution
    max_angle_percentile : float
        The larger percentile of one of the vectors to pick from the angle
        distribution
    mask_out : array_like
        if not None, should be (m, n) boolean numpy array.
        This parameter ensures exclusion of non-masked areas from calculations.
        This is relevant because elements like blood, sharpie marker,
        white space, etc may throw off the normalization somewhat.

    Returns
    -------
    w : array_like
        A 3x3 matrix of stain column vectors

    Note
    ----
    All input pixels not otherwise excluded are used in the computation of the
    principal plane and the angle distribution.

    See Also
    --------
    histomicstk.preprocessing.color_deconvolution.color_deconvolution
    histomicstk.preprocessing.color_deconvolution.separate_stains_xu_snmf

    References
    ----------
    .. [#] Van Eycke, Y. R., Allard, J., Salmon, I., Debeir, O., &
           Decaestecker, C. (2017).  Image processing in digital pathology: an
           opportunity to solve inter-batch variability of immunohistochemical
           staining.  Scientific Reports, 7.
    .. [#] Macenko, M., Niethammer, M., Marron, J. S., Borland, D.,
           Woosley, J. T., Guan, X., ... & Thomas, N. E. (2009, June).
           A method for normalizing histology slides for quantitative analysis.
           In Biomedical Imaging: From Nano to Macro, 2009.  ISBI'09.
           IEEE International Symposium on (pp. 1107-1110). IEEE.

    """
    # Image matrix
    m = utils.convert_image_to_matrix(im_sda)

    # mask out irrelevant values
    if mask_out is not None:
        keep_mask = numpy.equal(mask_out[..., None], False)
        keep_mask = numpy.tile(keep_mask, (1, 1, 3))
        keep_mask = utils.convert_image_to_matrix(keep_mask)
        m = m[:, keep_mask.all(axis=0)]

    # get rid of NANs and infinities
    m = utils.exclude_nonfinite(m)

    # Principal components matrix
    pcs = linalg.get_principal_components(m)
    # Input pixels projected into the PCA plane
    proj = pcs.T[:-1].dot(m)
    # Pixels above the magnitude threshold
    filt = proj[:, linalg.magnitude(proj) > minimum_magnitude]
    # The "angles"
    angles = _get_angles(filt)

    # The stain vectors

    def get_percentile_vector(p):
        return pcs[:, :-1].dot(filt[:, argpercentile(angles, p)])

    min_v = get_percentile_vector(min_angle_percentile)
    max_v = get_percentile_vector(max_angle_percentile)

    # The stain matrix
    w = complement_stain_matrix(linalg.normalize(
        numpy.array([min_v, max_v]).T))
    return w


def _get_angles(m):
    """Take a 2xN matrix of vectors and return a length-N array of an.

    ... angle-like quantity.
    Since this is an internal function, we assume that the values
    result from PCA, and so the second element of the vectors captures
    secondary variation -- and thus is the one that takes on both
    positive and negative values.

    """
    m = linalg.normalize(m)
    # "Angle" towards +x from the +y axis
    return (1 - m[1]) * numpy.sign(m[0])


def argpercentile(arr, p):
    """Calculate index in arr of element nearest the pth percentile."""
    # Index corresponding to percentile
    i = int(p * arr.size + 0.5)
    return numpy.argpartition(arr, i)[i]

File: rust/separate_stains_macenko_pca.rs
use crate::complement_stain_matrix::complement_stain_matrix;
use crate::linalg::{get_principal_components, magnitude, normalize};
use crate::utils::{convert_image_to_matrix, exclude_nonfinite};
use ndarray::{Array1, Array2, Array3, ArrayView2, Axis, s};
use rayon::prelude::*; // Add this at the top

/// Compute the stain matrix for color deconvolution with the Macenko method.
pub fn separate_stains_macenko_pca(
    im_sda: Array3<f64>,
    minimum_magnitude: f64,
    min_angle_percentile: f64,
    max_angle_percentile: f64,
    mask_out: Option<&ArrayView2<bool>>,
) -> Array2<f64> {
    // Convert image to matrix
    let mut m = convert_image_to_matrix(im_sda);
    // Mask out irrelevant values
    if let Some(mask) = mask_out {
        // mask is H×W; convert it to a flat list of pixel-indices where mask==false
        let mask_indices: Vec<usize> = mask
            .iter()
            .enumerate()
            .filter(|&(_, &is_masked)| !is_masked)
            .map(|(flat_idx, _)| flat_idx)
            .collect();
        m = select_columns_parallel(&m, &mask_indices);
    }

    // Exclude NaNs and infinities
    m = exclude_nonfinite(m);

    // Principal components matrix
    let pcs = get_principal_components(&m);

    // Project input pixels into the PCA plane
    let proj = pcs.t().slice(s![..-1, ..]).dot(&m);
    drop(m);

    // Pixels above the magnitude threshold
    let mag = magnitude(&proj);
    let filt_indices: Vec<usize> = mag
        .iter()
        .enumerate()
        .filter(|(_, v)| **v > minimum_magnitude)
        .map(|(i, _)| i)
        .collect();
    drop(mag);
    let filt = select_columns_parallel(&proj, &filt_indices);
    drop(proj);
    // The "angles"
    let angles = get_angles(&filt);

    // The stain vectors
    let percentiles = [min_angle_percentile, max_angle_percentile];
    let vectors: Vec<Array1<f64>> = percentiles
        .par_iter()
        .map(|&p| get_percentile_vector(&pcs, &filt, &angles, p))
        .collect();
    let min_v = &vectors[0];
    let max_v = &vectors[1];
    drop(pcs);
    drop(filt);
    drop(angles);

    // The stain matrix
    let mut stains = ndarray::Array2::<f64>::zeros((min_v.len(), 2));
    stains
        .axis_iter_mut(Axis(0))
        .into_par_iter()
        .enumerate()
        .for_each(|(i, mut row)| {
            row[0] = min_v[i];
            row[1] = max_v[i];
        });
    complement_stain_matrix(&normalize(&stains))
}

/// faster, lower‐overhead version of select_columns
fn select_columns_parallel(arr: &Array2<f64>, indices: &[usize]) -> Array2<f64> {
    let nrows = arr.nrows();
    let ncols = indices.len();
    if ncols == 0 {
        return Array2::zeros((nrows, 0));
    }

    // build one flat Vec in parallel
    let new_data: Vec<f64> = (0..nrows * ncols)
        .into_par_iter()
        .map(|flat| {
            let row = flat / ncols;
            let col = indices[flat % ncols];
            arr[[row, col]]
        })
        .collect();

    Array2::from_shape_vec((nrows, ncols), new_data).expect("Shape conversion should not fail")
}

/// compute the same “angles” but without a full normalize() allocation
fn get_angles(mat: &Array2<f64>) -> Array1<f64> {
    let n = mat.ncols();
    let r0 = mat.row(0);
    let r1 = mat.row(1);
    let mut ang = Array1::zeros(n);
    for i in 0..n {
        let x = r0[i];
        let y = r1[i];
        let mag = (x * x + y * y).sqrt();
        ang[i] = if mag != 0.0 {
            let xn = x / mag;
            let yn = y / mag;
            (1.0 - yn) * xn.signum()
        } else {
            0.0
        };
    }
    ang
}

fn argpercentile(arr: &Array1<f64>, p: f64) -> usize {
    let size = arr.len();
    let i = ((p * size as f64 + 0.5).min((size - 1) as f64)) as usize;
    let mut indices: Vec<usize> = (0..size).collect();
    indices.select_nth_unstable_by(i, |&a, &b| arr[a].partial_cmp(&arr[b]).unwrap());
    indices[i]
}

fn get_percentile_vector(
    pcs: &Array2<f64>,
    filt: &Array2<f64>,
    angles: &Array1<f64>,
    percentile: f64,
) -> Array1<f64> {
    let idx = argpercentile(angles, percentile);
    pcs.slice(s![.., ..pcs.ncols() - 1]).dot(&filt.column(idx))
}
